datalake/src/main/java/com/sequenceiq/datalake/service/sdx/dr/SdxBackupRestoreService.java

```diff
@@ -757,4 +757,86 @@ private boolean isRunningState(DatalakeOperationStatus.State state) {
+     * s3a://my-bucket
+     * s3a://my-bucket/
+     *
+     * @param backupLocation The original backup location.
+     * @return The updated backup location.
+     */
+    private String appendBackupsToAWSRAZRootLocation(String backupLocation) {
+        String[] splitBySlashes = backupLocation.split("//", 2);
+        if (splitBySlashes.length <= 1) {
+            LOGGER.error("Invalid backup location (missing '//'): {}", backupLocation);
+            return backupLocation;
+        }
+        String pathAfterSlashes = splitBySlashes[1];
+        if (("").equals(pathAfterSlashes)) {
+            LOGGER.error("Backup locaiton has nothing after prefix: {}", backupLocation);
+            return backupLocation;
+        }
+        if (!pathAfterSlashes.contains("/")) {
+            LOGGER.info("Appending '/backups' to the backup location {} because it is an AWS RAZ root directory location", backupLocation);
+            return backupLocation + "/backups";
+        }
+        long countOfSlash = pathAfterSlashes.chars().filter(ch -> ch == '/').count();
+        if (countOfSlash == 1 && pathAfterSlashes.endsWith("/")) {
+            LOGGER.info("Appending 'backups' to the backup location {} because it is an AWS RAZ and directory location", backupLocation);
+            return backupLocation + "backups";
+        }
+        return backupLocation;
+    }
+
+    /**
+     * Append "/backups" or "backups" to root directory for Azure and RAZ. Below are the two types of root directories:
+     * abfs://test@mydatalake.dfs.core.windows.net
+     * abfs://test@mydatalake.dfs.core.windows.net/
+     *
+     * @param backupLocation The storage location.
+     * @return The updated backup location.
+     */
+    public static String appendBackupsToAzureRAZRootLocation(String backupLocation) {
+        String[] splitByAt = backupLocation.split("@", 2);
+        if (splitByAt.length <= 1) {
+            LOGGER.error("Invalid backup location (missing '@'): {}", backupLocation);
+            return backupLocation;
+        }
+        String pathAfterAt = splitByAt[1];
+        String[] splitByDomain = pathAfterAt.split(".dfs.core.windows.net", 2);
+        if (splitByDomain.length <= 1) {
+            LOGGER.error("Invalid backup location (missing domain): {}", backupLocation);
+            return backupLocation;
+        }
+        String pathAfterDomain = splitByDomain[1];
+        if (pathAfterDomain.length() == 0) {
+            LOGGER.info("Appending '/backups' to the backup location {} because it is an Azure RAZ root directory location", backupLocation);
+            return backupLocation + "/backups";
+        }
+        if (("/").equals(pathAfterDomain)) {
+            LOGGER.info("Appending 'backups' to the backup location {} because it is an Azure RAZ root directory location", backupLocation);
+            return backupLocation + "backups";
+        }
```

Comment:
this looks like to me that you are overusing `String#split`
why don't you validate everything with a simple regex (the first 2 if), and do something like
`StringUtils.appendIfMissing(StringUtils.removeEnd(loc, "/"), "/backups")`

Reply:
Hi @lacikaaa , Sergiy helped me come up with a way to do this without using regex and split (because we think they are not stable enough). Im not validating if it is a valid S3 or Azure location anymore using the new way since it is not the purpose of this function. Please let me know if the updated version looks good to you. Thank you!!